crates/ruff/src/rules/ruff/typing.rs

```diff
@@ -7,40 +7,6 @@ use ruff_python_ast::typing::parse_type_annotation;
 use ruff_python_semantic::SemanticModel;
 use ruff_python_stdlib::sys::is_known_standard_library;
 
-/// Custom iterator to collect all the `|` separated expressions in a PEP 604
-/// union type.
-struct PEP604UnionIterator<'a> {
-    stack: Vec<&'a Expr>,
-}
-
-impl<'a> PEP604UnionIterator<'a> {
-    fn new(expr: &'a Expr) -> Self {
-        Self { stack: vec![expr] }
-    }
-}
-
-impl<'a> Iterator for PEP604UnionIterator<'a> {
-    type Item = &'a Expr;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        while let Some(expr) = self.stack.pop() {
-            match expr {
-                Expr::BinOp(ast::ExprBinOp {
-                    left,
-                    op: Operator::BitOr,
-                    right,
-                    ..
-                }) => {
-                    self.stack.push(left);
-                    self.stack.push(right);
-                }
-                _ => return Some(expr),
-            }
-        }
-        None
-    }
-}
```

Comment:
I think we can remove this entirely and instead do this recursion via `contains_none` and `contains_any`.

Reply:
Yeah.